compiler/src/iree/compiler/Codegen/Dialect/Codegen/IR/IREECodegenDialect.cpp

```diff
@@ -58,28 +58,130 @@ IREECodegenDialect::verifyOperationAttribute(Operation *op,
   StringRef symbol = attribute.getName().strref();
   Attribute attr = attribute.getValue();
   // This function verifies the validity of a specific operation attribute.
-  // - If the attribute's name matches `kTuningDefaultSpecAttrName`, make
-  //   sure it contains a single named sequence op with name `__kernel_config`.
+  // - If the attribute's name matches kTuningSpecDefaultEntrypointAttrName
+  // (`iree_codegen.tuning_spec_with_default_entrypoint`):
+  //   1. Ensure that the module contains a single named sequence operation with
+  //   the name `__kernel_config`.
+  //   2. Verify that this `__kernel_config` named sequence operation has the
+  //   attribute `iree_codegen.tuning_spec_entrypoint`.
+  //   3. Ensure that the named sequence operation contains exactly **one**
+  //   `ForeachMatchOp`.
+  //      - ForeachMatchOp must not have `flatten_results` and `restrict_root`
+  //        attributes.
+  //      - ForeachMatchOp must have exactly one argument of type any_op.
+  //      - ForeachMatchOp must have exactly one result of type any_op.
+  //   4. Ensure that only one named sequence operation with the
+  //   `iree_codegen.tuning_spec_entrypoint` attribute.
   // - If the attribute's name matches `kTuningSpecEntrypointAttrName`
-  // ("iree_codegen.tuning_spec_entrypoint"):
+  // (`iree_codegen.tuning_spec_entrypoint`):
   //   1. The attribute value must be a UnitAttr.
   //   2. If the operation is a transform::NamedSequenceOp:
   //      - The operation's function signature must satisfy the following:
-  //         a. It must have exactly one result type, and the result must be of
-  //         type `transform::AnyOpType`.
+  //         a. It must have exactly one result type, and the result must be
+  //         of type `transform::AnyOpType`.
   //         b. It must have exactly one argument type, and the argument must be
   //         of type `transform::AnyOpType`.
 
   if (symbol == kTuningSpecDefaultEntrypointAttrName) {
     if (auto moduleOp = dyn_cast<ModuleOp>(op)) {
-      if (!llvm::any_of(moduleOp.getOps<transform::NamedSequenceOp>(),
-                        [](transform::NamedSequenceOp op) {
-                          return op.getName() == kKernelConfigSpecName;
-                        })) {
+      transform::NamedSequenceOp kernelConfigOp;
+      int numTuningEntryPoints = 0;
+      for (Region &region : moduleOp->getRegions()) {
+        for (Block &block : region) {
+          for (Operation &op : block) {
+            if (auto namedSeqOp = dyn_cast<transform::NamedSequenceOp>(&op)) {
+              if (namedSeqOp.getName() == kKernelConfigSpecName) {
```

Comment:
Not what this logic became quite complex, I think we should pull this into a separate function with helper function that can check each condition. For example, one function can check that's there's exactly one default entrypoint while another one can check its contents.